
# Udemy Courses Analysis Project

## Project Overview

In this project, we will delve into a comprehensive analysis of a dataset containing information about Udemy courses. As an enthusiastic data analyst eager to work in the data field, you have a unique opportunity to gain valuable insights from this dataset while honing your skills in Python, SQL, Excel, and data analysis.

## Project Objective

The primary objective of this project is to extract actionable insights from the Udemy dataset. By leveraging your data analysis skills and industry expertise, we aim to achieve the following:

1.  **Gain Insights into Course Popularity:** We will explore which courses attract the most subscribers and why. This analysis will involve examining factors such as course subject, pricing, and content duration.
    
2.  **Pricing Strategies:** We will investigate Udemy's pricing strategies by analyzing the distribution of course prices. This analysis will help us understand how price affects course enrollment.
    
3.  **Subject Analysis:** We will categorize courses by subject and examine the distribution of courses within each subject. Are certain subjects more popular than others?
    
4.  **Trends and Patterns:** Identifying trends and patterns in the data will be a key focus. We will look for correlations between variables like the number of subscribers, the number of reviews, and course difficulty level.
    
5.  **Data Visualization:** Data visualization will play a crucial role in presenting our findings. We will use various charts and graphs to make the analysis results more accessible and understandable.
    
6.  **Python Code Optimization:** Throughout the project, we will ensure that our Python code adheres to the PEP-8 guidelines for code style and optimization. Clean and efficient code is essential for effective data analysis.
    
7.  **Documentation:** Detailed documentation of the code and analysis steps will be maintained, including comments in Python code. This documentation will serve as a reference for future work and collaborations.
    

## Dataset Description

The dataset consists of the following columns:

1.  `course_id`: Unique identifier for each course.
2.  `course_title`: The title or name of the course.
3.  `is_paid`: Indicates whether the course is paid (True/False).
4.  `price`: The price of the course.
5.  `num_subscribers`: The number of subscribers to the course.
6.  `num_reviews`: The number of reviews for the course.
7.  `num_lectures`: The number of lectures in the course.
8.  `level`: The difficulty level of the course (e.g., beginner, intermediate, advanced).
9.  `content_duration`: The duration of the course content.
10.  `published_timestamp`: The timestamp when the course was published.
11.  `subject`: The subject or category of the course.

## Import all required Libraries

In [18]:
import pandas as pd                         # For data manipulation and analysis 
import matplotlib.pyplot as plt             # For data visualization 
import seaborn as sns                       # For better data  visualization
import re                                   # For regular expression operations 

#### Read the dataset with the pandas library

In [36]:
df = pd.read_csv("7. Udemy Courses.csv", encoding="UTF-8")

#### Let's Explore the dataset

In [37]:
# Print the top 5 lines of the  file.
df.head(5)

course_id                                       course_title  is_paid  \
0     288942  #1 Piano Hand Coordination: Play 10th Ballad i...     True   
1    1170074  #10 Hand Coordination - Transfer Chord Ballad ...     True   
2    1193886  #12 Hand Coordination: Let your Hands dance wi...     True   
3    1116700  #4 Piano Hand Coordination: Fun Piano Runs in ...     True   
4    1120410  #5  Piano Hand Coordination:  Piano Runs in 2 ...     True   

  price  num_subscribers  num_reviews  num_lectures               level  \
0    35             3137           18            68          All Levels   
1    75             1593            1            41  Intermediate Level   
2    75              482            1            47  Intermediate Level   
3    75              850            3            43  Intermediate Level   
4    75              940            3            32  Intermediate Level   

  content_duration   published_timestamp              subject  
0        1.5 hours  2014-09-18T05:07:05Z  Musical Instruments  
1           1 hour  2017-04-12T19:06:34Z  Musical Instruments  
2        1.5 hours  2017-04-26T18:34:57Z  Musical Instruments  
3           1 hour  2017-02-21T23:48:18Z  Musical Instruments  
4          37 mins  2017-02-21T23:44:49Z  Musical Instruments

In [39]:
# Print the bottom 5 lines of the file.
df.tail(5)

course_id                                       course_title  is_paid  \
3677     328960  Your Own Site in 45 Min: The Complete Wordpres...     True   
3678     552700    Your Second Course on Piano: Two Handed Playing     True   
3679     631754  Zend Framework 2: Learn the PHP framework ZF2 ...     True   
3680    1225194  Zoho Books Gestion Financière d'Entreprise pas...    False   
3681     964478                Zombie Apocalypse Photoshop Actions     True   

     price  num_subscribers  num_reviews  num_lectures           level  \
3677   120             1566           29            36      All Levels   
3678    70             1018           12            22  Beginner Level   
3679    40              723          130            37      All Levels   
3680  Free              229            0            33      All Levels   
3681    50               12            1            15      All Levels   

     content_duration   published_timestamp              subject  
3677          4 hours  2015-04-20T22:15:17Z      Web Development  
3678          5 hours  2015-10-26T20:04:21Z  Musical Instruments  
3679        6.5 hours  2015-11-11T18:55:45Z      Web Development  
3680          2 hours  2017-05-26T16:45:55Z     Business Finance  
3681        1.5 hours  2016-09-26T22:19:48Z       Graphic Design

We can observe that, in the dataset, the courses with no price is denoted by "Free". This will be changed to 0 for better analysis.

In [40]:
# Change the Free price value with 0 
df["price"] = df["price"].apply(lambda x: 0 if x == "Free" else x)

df.tail(5)

course_id                                       course_title  is_paid  \
3677     328960  Your Own Site in 45 Min: The Complete Wordpres...     True   
3678     552700    Your Second Course on Piano: Two Handed Playing     True   
3679     631754  Zend Framework 2: Learn the PHP framework ZF2 ...     True   
3680    1225194  Zoho Books Gestion Financière d'Entreprise pas...    False   
3681     964478                Zombie Apocalypse Photoshop Actions     True   

     price  num_subscribers  num_reviews  num_lectures           level  \
3677   120             1566           29            36      All Levels   
3678    70             1018           12            22  Beginner Level   
3679    40              723          130            37      All Levels   
3680     0              229            0            33      All Levels   
3681    50               12            1            15      All Levels   

     content_duration   published_timestamp              subject  
3677          4 hours  2015-04-20T22:15:17Z      Web Development  
3678          5 hours  2015-10-26T20:04:21Z  Musical Instruments  
3679        6.5 hours  2015-11-11T18:55:45Z      Web Development  
3680          2 hours  2017-05-26T16:45:55Z     Business Finance  
3681        1.5 hours  2016-09-26T22:19:48Z       Graphic Design

Also, is_paid column is of boolean type. We will convert it to 0 and 1 for better analysis.

In [41]:
# Also, is_paid column is of boolean type. We will convert it to 0 and 1 for better analysis.
df["is_paid"] = df["is_paid"].apply(lambda x: 1 if x == True else 0)

df.tail(5)

course_id                                       course_title  is_paid  \
3677     328960  Your Own Site in 45 Min: The Complete Wordpres...        1   
3678     552700    Your Second Course on Piano: Two Handed Playing        1   
3679     631754  Zend Framework 2: Learn the PHP framework ZF2 ...        1   
3680    1225194  Zoho Books Gestion Financière d'Entreprise pas...        0   
3681     964478                Zombie Apocalypse Photoshop Actions        1   

     price  num_subscribers  num_reviews  num_lectures           level  \
3677   120             1566           29            36      All Levels   
3678    70             1018           12            22  Beginner Level   
3679    40              723          130            37      All Levels   
3680     0              229            0            33      All Levels   
3681    50               12            1            15      All Levels   

     content_duration   published_timestamp              subject  
3677          4 hours  2015-04-20T22:15:17Z      Web Development  
3678          5 hours  2015-10-26T20:04:21Z  Musical Instruments  
3679        6.5 hours  2015-11-11T18:55:45Z      Web Development  
3680          2 hours  2017-05-26T16:45:55Z     Business Finance  
3681        1.5 hours  2016-09-26T22:19:48Z       Graphic Design

Let's convert the content_duration column to minutes for better analysis.

In [47]:
# Convert time values to minutes and create a new column

def convert_to_minutes(time_str):
    hours = 0
    minutes = 0
    if "hour" in time_str:
        hours_match = re.search(r"(\d+(\.\d+)?)\s*hour", time_str)
        if hours_match:
            hours = float(hours_match.group(1))
    if "min" in time_str:
        minutes_match = re.search(r"(\d+)\s*min", time_str)
        if minutes_match:
            minutes = int(minutes_match.group(1))
    return int(hours * 60 + minutes)


df["course_duration (in minutes)"] = df["content_duration"].apply(
    convert_to_minutes
)



# Let's drop the  content_duration column
df.drop(columns=["content_duration"], inplace=True)
df.head(3)

NameError: name 're' is not defined

In [42]:
# Print the number of rows and columns in the data frame.
df.shape

(3682, 11)

So we have 3682 rows and 11 columns in our dataset. Let's check the data types of each column.

In [46]:
# Check the data types of the columns.
df.dtypes

course_id               int64
course_title           object
is_paid                 int64
price                   int32
num_subscribers         int64
num_reviews             int64
num_lectures            int64
level                  object
content_duration       object
published_timestamp    object
subject                object
dtype: object

We can observe that, some of the columns have incorrect data types. We will fix this in the next step.

Converted `price` to a numeric data type (float) for numerical analysis and calculations, enabling easier pricing strategy exploration. <br />

Changed `content_duration` to a float data type for improved analysis, facilitating better understanding of course duration trends.<br />

Transformed `published_timestamp` into a datetime data type, enabling time-based exploration and deeper insights into course publication patterns.<br />

In [43]:
# Change the dara types of the columns.

df["price"] = df["price"].astype("int")
df["content_duration"] = df["content_duration"].astype("int")
df["published_timestamp"] = pd.to_datetime(df["published_timestamp"])

ValueError: invalid literal for int() with base 10: '1.5 hours'

Check for null values in the data set

In [22]:
df.isnull().sum()

course_id              0
course_title           0
is_paid                0
price                  0
num_subscribers        0
num_reviews            0
num_lectures           0
level                  0
content_duration       0
published_timestamp    0
subject                0
dtype: int64

So, there are no null values in our dataset which is kind of a good thing. :)